docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,119 @@
+
+The controller publishes its certificate using [ZControllerCert](../api/APIv2.md#controller-certificates) and the device publishes its certificate using [ZEveCert](../api/APIv2.md#attestation), as described in the [API](../api/APIv2.md#messages) specification.
+
+## Sender authentication
+
+ECDH does not provide sender authentication. But the certificates used for ECDH are sent over the secure channel between the peers (which consists of transport security between the TLS endpoints and the object signing providing integrity and sender authentication between the microservice endpoints). That the sender authentication and integrity means that an attacker can not replace the certificates used for ECDH.
+
+(Perfect) Forward Secrecy is provided by having the controller periodically generate a new ECC certificate.
+
+## API components
+
+In addition to the above mechanisms for the controller and device to send their ECC certificates, there are several specific pieces in the EVE API to carry the state needed for encryption and re-keying.
+
+### Cipher Contexts
+
+The [Cipher Context](../api/proto/config/acipherinfo.proto#L26) is a construct used for defining hash algorithms, key exchange schemes like ECDH, etc., encryption schemes like AES256, etc., and certificate hashes used for the encryption by the controller.
+
+### Cipher Block
+
+The [Cipher block](../api/proto/config/acipherinfo.proto#L47) is a construct which has a cipher context id, [IV](https://en.wikipedia.org/wiki/Initialization_vector), encrypted data and hash of clear text (used for verification after decryption). Cipher block is a part of the object configuration which has encrypted data.
+
+### Encryption Block
+
+The [Encryption block](../api/proto/config/acipherinfo.proto#L66) is a construct used for encryption by the controller. The controller fills the encryption block with sensitive information and then encrypts it and provides encrypted data in the cipher block.
+
+## EVE packages
+
+### Cipher Package
+
+The [cipher](../pkg/pillar/cipher/handlecipher.go) package is called by the agents which want to decrypt the sensitive information. Caller provides cipher block and subscription contexts of controller certificates and cipher contexts.
+
+The cipher packet takes the cipher context id from the cipher block and then subscribes for that cipher context to the zedagent. Cipher context has the hashes of all the certificates which are required for the decryption.
+
+It implements the crypto operations required to retrieve the shared key from the controller, and also interacts with TPM (if present) for ECDH operations and returns an encryption block to the caller after decryption.
+
+## EVE agents
+
+Following are the key agents in EVE, which are involved in implementing this solution:
+
+### TPM Manager
+
+TPM Manager creates ECDH certificates using the TPM, and also implements the glue layer to talk to [go-tpm](https://github.com/google/go-tpm) for ECDH APIs.
```

Comment:
What about devices without tpm? How do they generate the certificate, and where do they store the key?

Reply:
@cshari-zededa how will things work when there is no TPM? We should add some wording to this document.

Reply:
@deitch I got text from Hari. Added to document.